src/coreclr/jit/valuenum.cpp

```diff
@@ -3460,39 +3421,117 @@ ValueNum ValueNumStore::EvalCastForConstantArgs(var_types typ, VNFunc func, Valu
-//                 EvalOpspecialized or EvalComparison
-//
-bool ValueNumStore::CanEvalForConstantArgs(VNFunc vnf)
+ValueNum ValueNumStore::EvalBitCastForConstantArgs(var_types dstType, ValueNum arg0VN)
 {
-    if (vnf < VNF_Boundary)
+    // Handles - when generating relocatable code - don't represent their final
+    // values, so we'll not fold bitcasts from them (always, for simplicity).
+    assert(!IsVNHandle(arg0VN));
+
+    var_types srcType = TypeOfVN(arg0VN);
+    assert((genTypeSize(srcType) == genTypeSize(dstType)) || (varTypeIsSmall(dstType) && (srcType == TYP_INT)));
+
+    union {
+        int           Int;
+        int64_t       Long;
+        target_size_t IntPtr;
+        float         Float;
+        double        Double;
+    } srcValue{};
+
+    switch (srcType)
     {
-        genTreeOps oper = genTreeOps(vnf);
+        case TYP_INT:
+            srcValue.Int = ConstantValue<int>(arg0VN);
+            break;
+        case TYP_LONG:
+            srcValue.Long = ConstantValue<int64_t>(arg0VN);
+            break;
+        case TYP_BYREF:
+            srcValue.IntPtr = ConstantValue<target_size_t>(arg0VN);
+            break;
+        case TYP_REF:
+            noway_assert(arg0VN == VNForNull());
+            srcValue.IntPtr = 0;
+            break;
+        case TYP_FLOAT:
+            srcValue.Float = ConstantValue<float>(arg0VN);
+            break;
+        case TYP_DOUBLE:
+            srcValue.Double = ConstantValue<double>(arg0VN);
+            break;
+        default:
+            unreached();
+    }
 
-        switch (oper)
-        {
-            // Only return true for the node kinds that have code that supports
-            // them in EvalOP, EvalOpspecialized or EvalComparison
+    // "BitCast<small type>" has the semantic of only changing the upper bits (without truncation).
+    if (varTypeIsSmall(dstType))
+    {
+        assert(FitsIn(varTypeToSigned(dstType), srcValue.Int) || FitsIn(varTypeToUnsigned(dstType), srcValue.Int));
+    }
 
-            // Unary Ops
-            case GT_NEG:
-            case GT_NOT:
-            case GT_BSWAP16:
-            case GT_BSWAP:
+    switch (dstType)
+    {
+        case TYP_BOOL:
+        case TYP_UBYTE:
+            return VNForIntCon(static_cast<uint8_t>(srcValue.Int));
+        case TYP_BYTE:
+            return VNForIntCon(static_cast<int8_t>(srcValue.Int));
+        case TYP_USHORT:
+            return VNForIntCon(static_cast<uint16_t>(srcValue.Int));
+        case TYP_SHORT:
+            return VNForIntCon(static_cast<int16_t>(srcValue.Int));
+        case TYP_INT:
+            return VNForIntCon(srcValue.Int);
+        case TYP_LONG:
+            return VNForLongCon(srcValue.Long);
+        case TYP_BYREF:
+            return VNForByrefCon(srcValue.IntPtr);
+        case TYP_FLOAT:
+            return VNForFloatCon(srcValue.Float);
+        case TYP_DOUBLE:
+            return VNForDoubleCon(srcValue.Double);
+        default:
+            unreached();
+    }
```

Comment:
This kind of reinterpretation is undefined behavior in C++. We've recently had an issue because of similar kind of reinterpretation (#67669). While it may be ok for the compilers we use, I would still feel better if this was changed to `memcpy` into and out of a common byte buffer.